AX: need to handle dynamically-created and nested aria-modal dialogs

minorninth opened the pull request:
#### 77d74fdb4d16d5bcf79cf7a270b7e2dcbaf3799f
<pre>
AX: need to handle dynamically-created and nested aria-modal dialogs
<a href="https://bugs.webkit.org/show_bug.cgi?id=280643">https://bugs.webkit.org/show_bug.cgi?id=280643</a>
<a href="https://rdar.apple.com/137000722">rdar://137000722</a>

Reviewed by Tyler Wilcock.

In isolated tree mode, WebKit wasn&apos;t properly handling the case where
an aria-modal dialog is dynamically created and added to the DOM.  In
addition, WebKit didn&apos;t have any logic to pick the deepest aria-modal
dialog if multiple dialogs are nested. Adding this logic matches
Chrome&apos;s behavior now.  In both cases, the issue was correctly
ignoring everything outside of the dialog while focus is inside.

* LayoutTests/accessibility/mac/dynamic-modal-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-modal.html: Added.
* LayoutTests/accessibility/mac/nested-modal-expected.txt: Added.
* LayoutTests/accessibility/mac/nested-modal.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateCurrentModalNode):
(WebCore::AXObjectCache::handleAttributeChange):

Canonical link: <a href="https://commits.webkit.org/284613@main">https://commits.webkit.org/284613@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d330ce3d2e6e5ccc97cf400eabb8f28a1249e75f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69666 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49066 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22416 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73751 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20824 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71783 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56867 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20675 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55368 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13828 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72732 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44739 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60091 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35847 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17529 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19201 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63341 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17874 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75464 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13648 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17129 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63037 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13688 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60174 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62970 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15527 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10979 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4567 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44870 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45944 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47208 "Build was cancelled. Recent messages:OS: Sonoma (14.5), Xcode: 15.4") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45685 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -380,10 +380,11 @@ void AXObjectCache::updateCurrentModalNode(WillRecomputeFocus willRecomputeFocus
         SetForScope retrievingCurrentModalNode(m_isRetrievingCurrentModalNode, true);
         // If any of the modal nodes contains the keyboard focus, we want to pick that one.
+        // If multiple contain the keyboard focus, we want the deepest.
         // If not, we want to pick the last visible dialog in the DOM.
```
This "If not" comment makes less sense now that it doesn't immediately follow the first comment line. Maybe it should be updated to say something like "If no modal contains focus, we want to pick ..."?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -2705,8 +2717,10 @@ void AXObjectCache::handleAttributeChange(Element* element, const QualifiedName&
         postNotification(element, AXInvalidStatusChanged);
-    else if (attrName == aria_modalAttr)
+    else if (attrName == aria_modalAttr) {
+        m_modalNodesInitialized = false;
```
Setting this to `false` means we'll need to do a full DOM walk. I see that we set this to false too when `aria-hidden` changes above the active modal, so this PR matches the behavior there.

But I wonder if it would make sense to just update `m_modalElements` directly here rather than set `m_modalNodesInitialized = false`? Wouldn't it be something like:

```cpp
else if (attrName == aria_modalAttr) {
    // aria-modal changed, so the element may have become modal or un-modal.
    if (isModalElement(*element))
        m_modalElements.appendIfNotContains(element);
    else
        m_modalElements.removeAll(element);

    deferModalChange(*element);
}
```

But maybe I'm missing some nuance here, or maybe it'd be preferable to take the potentially simpler but less efficient approach of setting `m_modalNodesInitialized = false`. What is your opinion?

minorninth replied:
I think you're right. When aria-hidden changes, it could potentially affect a whole subtree, but if aria-modal changes then I think it's safe to just check this element.

twilco commented on LayoutTests/accessibility/mac/nested-modal.html:
```diff
@@ -0,0 +1,80 @@
+    </div>
+  </div>
+
+<script>
```
Our test style typically doesn't indent a level inside the script tag, helps avoid a bit of rightward drift. Ditto for the other test.

twilco commented on LayoutTests/accessibility/mac/nested-modal.html:
```diff
@@ -0,0 +1,80 @@
+                return focus && focus.domIdentifier;
+            });
+
+
```
Extra newline here I think.

twilco commented on LayoutTests/accessibility/mac/nested-modal.html:
```diff
@@ -0,0 +1,80 @@
+            testOutput += "Initial focus: " + focus.domIdentifier + "\n";
+
+            before = accessibilityController.accessibleElementById("before");
+            testOutput += "Before button reachable: " + (before != null) + "\n";
```
Take it or leave it: Quite a few lines in this test do string concatenation using:

```
testOutput += "foo " + expression + " bar";
```

But in other tests, we've tended towards using format strings:

```
testOutput += `foo ${expression} bar`;
```

I don't feel too strongly though.

 Ditto for the other test.

twilco reviewed (commented):
Thanks very much for taking this on!

twilco reviewed (approved):
Thanks @minorninth!

minorninth committed:
AX: need to handle dynamically-created and nested aria-modal dialogs
https://bugs.webkit.org/show_bug.cgi?id=280643
rdar://137000722

Reviewed by Tyler Wilcock.

In isolated tree mode, WebKit wasn't properly handling the case where
an aria-modal dialog is dynamically created and added to the DOM.  In
addition, WebKit didn't have any logic to pick the deepest aria-modal
dialog if multiple dialogs are nested. Adding this logic matches
Chrome's behavior now.  In both cases, the issue was correctly
ignoring everything outside of the dialog while focus is inside.

* LayoutTests/accessibility/mac/dynamic-modal-expected.txt: Added.
* LayoutTests/accessibility/mac/dynamic-modal.html: Added.
* LayoutTests/accessibility/mac/nested-modal-expected.txt: Added.
* LayoutTests/accessibility/mac/nested-modal.html: Added.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateCurrentModalNode):
(WebCore::AXObjectCache::handleAttributeChange):

Canonical link: https://commits.webkit.org/284613@main
